raft.go

```diff
@@ -627,33 +630,41 @@ func (r *Raft) leaderLoop() {
 				}
 			}
 
-			var numProcessed int
 			start := time.Now()
+			var groupReady []*list.Element
+			var groupFutures = make(map[uint64]*logFuture)
+			var lastInGroup uint64
 
-			for {
-				e := r.leaderState.inflight.Front()
-				if e == nil {
-					break
-				}
+			// Pull all inflight logs that are committed off the queue.
+			for e := r.leaderState.inflight.Front(); e != nil; e = e.Next() {
 				commitLog := e.Value.(*logFuture)
 				idx := commitLog.log.Index
 				if idx > commitIndex {
+					// Don't go past the committed index
 					break
 				}
+
 				// Measure the commit time
 				metrics.MeasureSince([]string{"raft", "commitTime"}, commitLog.dispatch)
+				groupReady = append(groupReady, e)
+				groupFutures[idx] = commitLog
+				lastInGroup = idx
+			}
 
-				r.processLogs(idx, commitLog)
+			// Process the group
+			if len(groupReady) != 0 {
+				r.processLogs(lastInGroup, groupFutures)
 
-				r.leaderState.inflight.Remove(e)
-				numProcessed++
+				for _, e := range groupReady {
+					r.leaderState.inflight.Remove(e)
+				}
 			}
 
 			// Measure the time to enqueue batch of logs for FSM to apply
 			metrics.MeasureSince([]string{"raft", "fsm", "enqueue"}, start)
 
 			// Count the number of logs enqueued
-			metrics.SetGauge([]string{"raft", "commitNumLogs"}, float32(numProcessed))
+			metrics.SetGauge([]string{"raft", "commitNumLogs"}, float32(len(groupReady)))
```

Comment:
🤔 I guess this is where you copied the Gauge thing from. I just don't think this does anything useful. I guess our existing stats for this are essentially meaningless too...

Reply:
We should consider updating this in a separate PR